pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 			Name:      "distributor_forward_samples_total",
 			Help:      "The total number of samples the Distributor forwarded.",
 		}),
+		exemplarsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: "cortex",
+			Name:      "distributor_forward_exemplars_total",
+			Help:      "The total number of exemplars the Distributor forwarded.",
+		}),
 		requestLatencyHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
 			Namespace: "cortex",
 			Name:      "distributor_forward_requests_latency_seconds",
 			Help:      "The client-side latency of requests to forward metrics made by the Distributor.",
 			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 30},
 		}),
 	}
+
+	f.workerWg.Add(f.cfg.RequestConcurrency)
+	for i := 0; i < f.cfg.RequestConcurrency; i++ {
+		go f.worker()
+	}
+
+	return f
 }
 
-func (r *forwarder) NewRequest(ctx context.Context, tenant string, rules validation.ForwardingRules) Request {
-	return &request{
-		ctx:    ctx,
-		client: &r.client, // http client should be re-used so open connections get re-used.
-		pools:  &r.pools,
+func (f *forwarder) Stop() {
+	close(f.reqCh)
+	f.workerWg.Wait()
+}
 
-		tsByEndpoint: make(map[string]*[]mimirpb.PreallocTimeseries),
+func (f *forwarder) Forward(ctx context.Context, rules validation.ForwardingRules, in []mimirpb.PreallocTimeseries) (TimeseriesCounts, []mimirpb.PreallocTimeseries, chan error) {
+	if !f.cfg.Enabled {
+		errCh := make(chan error)
+		close(errCh)
+		return TimeseriesCounts{}, in, errCh
+	}
 
-		rules:           rules,
-		timeout:         r.cfg.RequestTimeout,
-		propagateErrors: r.cfg.PropagateErrors,
+	notIngestedCounts, toIngest, tsByTargets := f.splitByTargets(in, rules)
+	defer f.pools.putTsByTargets(tsByTargets)
 
-		requests: r.requestsTotal,
-		errors:   r.errorsTotal,
-		samples:  r.samplesTotal,
-		latency:  r.requestLatencyHistogram,
+	var requestWg sync.WaitGroup
```

Comment:
If I'm reading this code right:
* `forwarder` creates a pool of worker goroutines
* We're submitting jobs for them to run every time `Forward` is called and giving the caller a channel to read errors.

Seems like a job pool might be something we already have in some utility code or dskit. Offhand, there's [this method](https://github.com/grafana/dskit/blob/main/concurrency/runner.go#L87-L115) but it doesn't enforce a max number of goroutines across multiple calls to `Forward` at the same time.

Reply:
> If I'm reading this code right:
>
> * forwarder creates a pool of worker goroutines
> * We're submitting jobs for them to run every time Forward is called and giving the caller a channel to read errors.

Correct

> Seems like a job pool might be something we already have in some utility code or dskit. Offhand, there's [this method](https://github.com/grafana/dskit/blob/main/concurrency/runner.go#L87-L115) but it doesn't enforce a max number of goroutines across multiple calls to Forward at the same time.

You're right that this might likely be a pattern which is repeated somewhere else, so it would probably make sense to have support for this in dskit. Since we're under time pressure to get this PR merged I think I shouldn't add the support to dskit now and then update dskit in this PR because then there's a risk about further delays in the dskit PR, but I'd be happy to later implement a generic way for this worker pool pattern in dskit and migrate the forwarder to use it.

We could solve this by using dskit's existing `ForEachJob()`, but as you already mentioned this would mean that there won't be any concurrency limit on the forwarding requests which could lead to unbounded memory growth if a forwarding target is slow, plus this code is also quite sensitive to allocations because we're expecting to ship rather large amounts of data to forwarding targets and having a static pool of workers should lead to less allocations.

Reply:
:+1: fine to do this later